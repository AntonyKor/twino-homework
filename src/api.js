export async function getQuestions() {
  const resp = await fetch("/suitability");
  const data = await resp.json();

  return data.groups;
}
