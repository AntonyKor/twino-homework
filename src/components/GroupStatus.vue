<template>
  <div class="group-status">
    <h2>{{ groupName }}</h2>
    <p>{{ Math.round(groupProgress) }}% completed</p>
  </div>
</template>

<script>
export default {
  props: {
    groupProgress: {
      type: Number,
      required: true,
    },
    groupName: {
      type: String,
      default: "",
    },
  },
};
</script>

<style scoped lang="scss">
.group-status {
  margin-top: 30px;
  text-align: center;

  h2 {
    font-size: 1rem;

    @media screen and (min-width: 360px) {
      margin: 0 20px 0 0;
    }
  }

  p {
    margin: 0;
    color: #9e9fa2;
    font-size: 1.3rem;
  }

  @media screen and (min-width: 360px) {
    display: flex;
    justify-content: space-between;
    align-items: center;
  }
}
</style>
