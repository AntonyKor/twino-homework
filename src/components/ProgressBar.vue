<template>
  <div class="progress-bar" :style="barStyles">
    <div
      v-for="n in groupsNumber"
      :key="n"
      :class="{
        filled: n - 1 < currentGroup,
        active: n - 1 === currentGroup,
      }"
      class="group"
    >
      <div
        v-if="n - 1 === currentGroup"
        :style="currentGroupProgressStyles"
        class="filled"
      ></div>
    </div>
  </div>
</template>

<script>
export default {
  props: {
    groupsNumber: {
      type: Number,
      required: true,
    },
    currentGroup: {
      type: Number,
      required: true,
    },
    currentGroupProgress: {
      type: Number,
      required: true,
    },
  },

  computed: {
    barStyles() {
      return { gridTemplateColumns: `repeat(${this.groupsNumber}, 1fr)` };
    },

    currentGroupProgressStyles() {
      return { width: `${this.currentGroupProgress}%` };
    },
  },
};
</script>

<style scoped lang="scss">
.progress-bar {
  display: grid;
  grid-gap: 12px;
}

.group {
  background-color: #eef1f8;
  height: 6px;
}

.filled {
  background-color: #81c784;
}

.active {
  position: relative;
  background-color: #d9eeda;

  & > div {
    position: absolute;
    top: 0;
    bottom: 0;
    left: 0;
  }
}
</style>
