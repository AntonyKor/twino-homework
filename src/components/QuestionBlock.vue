<template>
  <div>
    <h3>{{ question.question }}</h3>
    <p v-for="(option, index) in question.answers" :key="option">
      <label>
        <BaseRadioButton v-model="answer" :value="index" name="answer" />
        {{ option }}
      </label>
    </p>
  </div>
</template>

<script>
import BaseRadioButton from "./BaseRadioButton.vue";

export default {
  components: {
    BaseRadioButton,
  },

  props: {
    question: {
      type: Object,
      required: true,
    },
  },

  computed: {
    answer: {
      get() {
        return this.question.answer;
      },
      set(answer) {
        this.$emit("set-answer", answer);
      },
    },
  },
};
</script>

<style lang="scss" scoped>
label {
  display: flex;
  align-items: center;
}

div {
  margin-bottom: 30px;
}
</style>
