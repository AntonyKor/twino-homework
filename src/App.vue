<template>
  <div id="app">
    <h1>Suitabitity test</h1>
    <ProgressBar
      :groupsNumber="groups.length"
      :currentGroup="currentGroupIndex"
      :currentGroupProgress="currentGroupProgress"
    />
    <template v-if="!isOverviewShown">
      <GroupStatus
        v-if="groups[openedGroupIndex]"
        :groupProgress="openedGroupProgress"
        :groupName="groups[openedGroupIndex].title"
      />
      <QuestionBlock
        v-if="openedQuestion"
        :question="openedQuestion"
        @set-answer="setAnswer"
      />
      <p v-if="isErrorShown" class="error">Please provide answer</p>
      <button
        v-if="currentGroupIndex === groups.length"
        @click="isOverviewShown = true"
        class="next"
      >
        View overview
      </button>
      <button v-if="!isOpenedQuestionLast" @click="openNext" class="next">
        Next question
      </button>
      <button
        v-if="openedGroupIndex || openedQuestionIndex"
        @click="openPrevious"
        class="previous"
      >
        Previous question
      </button>
    </template>
    <OverviewScreen
      v-else
      :groups="groups"
      @question-select="handleQuestionClick"
    />
  </div>
</template>

<script>
import ProgressBar from "@/components/ProgressBar.vue";
import GroupStatus from "@/components/GroupStatus.vue";
import QuestionBlock from "@/components/QuestionBlock.vue";
import OverviewScreen from "@/components/OverviewScreen.vue";
import { getQuestions } from "@/api";

export default {
  components: {
    ProgressBar,
    GroupStatus,
    QuestionBlock,
    OverviewScreen,
  },

  async created() {
    this.groups = await getQuestions();
  },

  data() {
    return {
      groups: [],
      openedGroupIndex: 0,
      openedQuestionIndex: 0,
      isErrorShown: false,
      isOverviewShown: false,
    };
  },

  computed: {
    currentGroupIndex() {
      const index =
        this.groups.findIndex(
          (group) =>
            !group.questions.every((question) => question.answer !== undefined)
        ) ?? this.groups.length;

      return index === -1 ? this.groups.length : index;
    },
    currentGroupProgress() {
      return this.getGroupStatus(this.currentGroupIndex);
    },
    openedGroupProgress() {
      return this.getGroupStatus(this.openedGroupIndex);
    },
    openedQuestion() {
      return this.groups[this.openedGroupIndex]?.questions[
        this.openedQuestionIndex
      ];
    },
    isOpenedQuestionLast() {
      return (
        this.groups.length - 1 === this.openedGroupIndex &&
        this.groups[this.openedGroupIndex].questions.length - 1 ===
          this.openedQuestionIndex
      );
    },
  },

  methods: {
    getGroupStatus(groupIndex) {
      if (!this.groups[groupIndex]) {
        return 0;
      }

      const { questions } = this.groups[groupIndex];
      const answeredQuestions = questions.reduce(
        (acc, question) => (question.answer !== undefined ? acc + 1 : acc),
        0
      );
      return (answeredQuestions / questions.length) * 100;
    },
    setAnswer(answer) {
      this.$set(
        this.groups[this.openedGroupIndex].questions[this.openedQuestionIndex],
        "answer",
        answer
      );
      this.isErrorShown = false;
    },
    openNext() {
      if (this.openedQuestion.answer === undefined) {
        this.isErrorShown = true;
        return;
      }

      this.openedQuestionIndex++;

      if (
        this.groups[this.openedGroupIndex]?.questions.length <=
        this.openedQuestionIndex
      ) {
        this.openedQuestionIndex = 0;
        this.openedGroupIndex++;
      }
    },
    openPrevious() {
      this.isErrorShown = false;
      this.openedQuestionIndex--;
      if (this.openedQuestionIndex < 0) {
        this.openedQuestionIndex = 0;

        this.openedGroupIndex = this.openedGroupIndex
          ? this.openedGroupIndex - 1
          : 0;
      }
    },
    handleQuestionClick(data) {
      this.openedGroupIndex = data.groupIndex;
      this.openedQuestionIndex = data.questionIndex;
      this.isOverviewShown = false;
    },
  },
};
</script>

<style lang="scss">
#app {
  max-width: 1090px;
  min-width: 240px;
  margin: 0 30px;
  font-family: Avenir, Helvetica, Arial, sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  color: #213560;

  h1 {
    font-size: 3rem;
  }

  .error {
    margin: 30px 0;
    padding: 10px;
    background-color: #fff6f4;
    border: 2px solid #fe513a;
    border-radius: 5px;
    text-align: center;
  }

  .next {
    display: block;
    margin-bottom: 20px;
    padding: 20px;
    width: 100%;
    background-color: #2e50b6;
    border: 0;
    border-radius: 5px;
    outline: none;
    color: #fff;
    font-size: 1.1rem;
    font-weight: bold;
  }

  .previous {
    display: block;
    margin-bottom: 20px;
    width: 100%;
    background: none;
    border: 0;
    outline: none;
    color: #0052b4;
    font-size: 1.1rem;
    text-decoration: underline;
  }
}
</style>
