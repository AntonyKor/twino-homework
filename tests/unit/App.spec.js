import { shallowMount } from "@vue/test-utils";
import Vue from "vue";
import App from "@/App.vue";
import ProgressBar from "@/components/ProgressBar.vue";
import GroupStatus from "@/components/GroupStatus.vue";
import QuestionBlock from "@/components/QuestionBlock.vue";
import { getQuestions } from "@/api";

jest.mock("@/api", () => ({
  getQuestions: jest.fn().mockResolvedValue([
    {
      title: "group1",
      questions: [
        { question: "question1", answers: ["a1", "a2"], answer: 1 },
        { question: "question2", answers: ["a3"] },
      ],
    },
    {
      title: "group2",
      questions: [{ question: "question3", answers: ["a5", "a6"] }],
    },
  ]),
}));

describe("App", () => {
  const renderAppWithData = async () => {
    const wrapper = shallowMount(App);

    await Vue.nextTick();
    await Vue.nextTick();

    return wrapper;
  };
  const goToLastPage = async (wrapper) => {
    await wrapper.find(".next").trigger("click");
    wrapper.findComponent(QuestionBlock).vm.$emit("set-answer", 0);
    await Vue.nextTick();
    await wrapper.find(".next").trigger("click");
  };

  it("should match snapshot before data was loaded", () => {
    const wrapper = shallowMount(App);

    expect(wrapper.element).toMatchSnapshot();
  });

  it("should match snapshot after data was loaded", async () => {
    const wrapper = await renderAppWithData();

    expect(wrapper.element).toMatchSnapshot();
  });

  it("should call getQuestions", () => {
    shallowMount(App);

    expect(getQuestions).toBeCalled();
  });

  it("should pass props to ProgressBar", async () => {
    const wrapper = await renderAppWithData();

    expect(wrapper.findComponent(ProgressBar).props()).toEqual({
      currentGroup: 0,
      currentGroupProgress: 50,
      groupsNumber: 2,
    });
  });

  it("should pass props to GroupStatus", async () => {
    const wrapper = await renderAppWithData();

    expect(wrapper.findComponent(GroupStatus).props()).toEqual({
      groupName: "group1",
      groupProgress: 50,
    });
  });

  it("should pass current question to QuestionBlock", async () => {
    const wrapper = await renderAppWithData();

    expect(wrapper.findComponent(QuestionBlock).props()).toEqual({
      question: { question: "question1", answers: ["a1", "a2"], answer: 1 },
    });
  });

  it("should change current question on click next button if question is answered", async () => {
    const wrapper = await renderAppWithData();

    await wrapper.find(".next").trigger("click");

    expect(wrapper.findComponent(QuestionBlock).props()).toEqual({
      question: { question: "question2", answers: ["a3"] },
    });
  });

  it("should not change current question on click next button if question is not answered", async () => {
    const wrapper = await renderAppWithData();
    await wrapper.find(".next").trigger("click");

    await wrapper.find(".next").trigger("click");

    expect(wrapper.find(".next")).toBeTruthy();
  });

  it("should not render error at start", async () => {
    const wrapper = await renderAppWithData();

    expect(wrapper.find(".error").exists()).toBeFalsy();
  });

  it("should render error after click next button if question is not answered", async () => {
    const wrapper = await renderAppWithData();
    await wrapper.find(".next").trigger("click");

    await wrapper.find(".next").trigger("click");

    expect(wrapper.find(".error").exists()).toBeTruthy();
  });

  it("should change ProgressBar props after answer", async () => {
    const wrapper = await renderAppWithData();
    await wrapper.find(".next").trigger("click");

    wrapper.findComponent(QuestionBlock).vm.$emit("set-answer", 2);
    await Vue.nextTick();

    expect(wrapper.findComponent(ProgressBar).props()).toEqual({
      currentGroup: 1,
      currentGroupProgress: 0,
      groupsNumber: 2,
    });
  });

  it("should change GroupStatus props after answer", async () => {
    const wrapper = await renderAppWithData();
    await wrapper.find(".next").trigger("click");

    wrapper.findComponent(QuestionBlock).vm.$emit("set-answer", 2);

    expect(wrapper.findComponent(GroupStatus).props()).toEqual({
      groupName: "group1",
      groupProgress: 100,
    });
  });

  it("should change GroupStatus props after navigating to next group", async () => {
    const wrapper = await renderAppWithData();
    await wrapper.find(".next").trigger("click");
    wrapper.findComponent(QuestionBlock).vm.$emit("set-answer", 2);

    await wrapper.find(".next").trigger("click");

    expect(wrapper.findComponent(GroupStatus).props()).toEqual({
      groupName: "group2",
      groupProgress: 0,
    });
  });

  it("should not render back button on first question", async () => {
    const wrapper = await renderAppWithData();

    expect(wrapper.find(".previous").exists()).toBeFalsy();
  });

  it("should render back button on other questions", async () => {
    const wrapper = await renderAppWithData();

    await wrapper.find(".next").trigger("click");

    expect(wrapper.find(".previous").exists()).toBeTruthy();
  });

  it("should change current question on click back button", async () => {
    const wrapper = await renderAppWithData();
    await wrapper.find(".next").trigger("click");

    await wrapper.find(".previous").trigger("click");

    expect(wrapper.findComponent(QuestionBlock).props()).toEqual({
      question: { question: "question1", answers: ["a1", "a2"], answer: 1 },
    });
  });

  it("should not render next button on last question", async () => {
    const wrapper = await renderAppWithData();

    await goToLastPage(wrapper);

    expect(wrapper.find(".next").exists()).toBeFalsy();
  });

  it("should not render view overview button if test is not passed", async () => {
    const wrapper = await renderAppWithData();

    await goToLastPage(wrapper);

    expect(wrapper.find(".next").exists()).toBeFalsy();
  });

  it("should render view overview button if test is passed", async () => {
    const wrapper = await renderAppWithData();

    await goToLastPage(wrapper);
    wrapper.findComponent(QuestionBlock).vm.$emit("set-answer", 0);
    await Vue.nextTick();

    expect(wrapper.find(".next").exists()).toBeTruthy();
    expect(wrapper.find(".next").text()).toEqual("View overview");
  });
});
