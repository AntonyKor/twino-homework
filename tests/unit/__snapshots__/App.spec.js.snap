// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot after data was loaded 1`] = `
<div
  id="app"
>
  <h1>
    Suitabitity test
  </h1>
   
  <progressbar-stub
    currentgroup="0"
    currentgroupprogress="50"
    groupsnumber="2"
  />
   
  <groupstatus-stub
    groupname="group1"
    groupprogress="50"
  />
   
  <questionblock-stub
    question="[object Object]"
  />
   
  <!---->
   
  <!---->
   
  <button
    class="next"
  >
    
      Next question
    
  </button>
   
  <!---->
</div>
`;

exports[`App should match snapshot before data was loaded 1`] = `
<div
  id="app"
>
  <h1>
    Suitabitity test
  </h1>
   
  <progressbar-stub
    currentgroup="0"
    currentgroupprogress="0"
    groupsnumber="0"
  />
   
  <!---->
   
  <!---->
   
  <!---->
   
  <button
    class="next"
  >
    
      View overview
    
  </button>
   
  <button
    class="next"
  >
    
      Next question
    
  </button>
   
  <!---->
</div>
`;
